### PR TITLE
[GEOS-8944] GWC fails to cache tiles if GeoFence with source IP driven rules are used for authentication

### PR DESCRIPTION
**Pull request by aaime**
Associated issue: https://osgeo-org.atlassian.net/browse/GEOS-8944

### CODE REVIEW

File: src/gwc/src/main/java/org/geoserver/gwc/FakeHttpServletRequest.java [hunk 1]
```diff
@@ -62,6 +66,13 @@ public FakeHttpServletRequest(
         this.parameterMap = parameterMap;
         this.cookies = cookies;
         this.workspace = workspace;
+        // grab the original request from Spring to forward security related attributes
+        // such as requests host, ports and headers
+        this.original =
+                Optional.ofNullable(
+                                (ServletRequestAttributes)
+                                        RequestContextHolder.getRequestAttributes())
+                        .map(atts -> atts.getRequest());
```
Comment (nmco) [inline src/gwc/src/main/java/org/geoserver/gwc/FakeHttpServletRequest.java]: Not sure if you notice, a method reference can be used here:
```
        this.original =
                Optional.ofNullable(
                                (ServletRequestAttributes)
                                        RequestContextHolder.getRequestAttributes())
                        .map(ServletRequestAttributes::getRequest);
```

File: src/gwc/src/main/java/org/geoserver/gwc/FakeHttpServletRequest.java [hunk 2]
```diff
@@ -300,35 +312,35 @@ public String getRealPath(String arg0) {
     }
 
     public String getRemoteAddr() {
-        return "127.0.0.1";
+        return original.map(r -> r.getRemoteAddr()).orElse("127.0.0.1");
```
Comment (nmco) [inline src/gwc/src/main/java/org/geoserver/gwc/FakeHttpServletRequest.java]: Wondering, is there a valid situation where the remote address will not be available ? I can't think of any. I would feel more safe throwing a  ``ServletDebugException``. The same opinion for the port (method above).

File: src/gwc/src/test/java/org/geoserver/gwc/HttpRequestRecorderCallback.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
+package org.geoserver.gwc;
```
Comment (nmco) [inline src/gwc/src/test/java/org/geoserver/gwc/HttpRequestRecorderCallback.java]: Missing header.

File: src/gwc/src/test/java/org/geoserver/gwc/HttpRequestRecorderCallback.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.http.HttpServletRequest;
+import org.geoserver.ows.AbstractDispatcherCallback;
+import org.geoserver.ows.Request;
+
+public class HttpRequestRecorderCallback extends AbstractDispatcherCallback {
```
Comment (nmco) [inline src/gwc/src/test/java/org/geoserver/gwc/HttpRequestRecorderCallback.java]: Could be made ``final``, I know this is a tests support class, but I would add some minor comment explaining what is going on:

```
Integration tests support class that keeps track of the requests send to the dispatcher.
```

**Comment by nmco** [comment]
Looks good, just some minor comments.

File: src/gwc/src/test/java/org/geoserver/gwc/HttpRequestRecorderCallback.java (hunk 3 above)
Comment (aaime) [reply src/gwc/src/test/java/org/geoserver/gwc/HttpRequestRecorderCallback.java]: Will add

File: src/gwc/src/main/java/org/geoserver/gwc/FakeHttpServletRequest.java (hunk 2 above)
Comment (aaime) [reply src/gwc/src/main/java/org/geoserver/gwc/FakeHttpServletRequest.java]: That might break existing code, better to preserve backwards compatibility

File: src/gwc/src/main/java/org/geoserver/gwc/FakeHttpServletRequest.java (hunk 1 above)
Comment (aaime) [reply src/gwc/src/main/java/org/geoserver/gwc/FakeHttpServletRequest.java]: I did not think about it, but not sure replacing 26 chars with 37 to do the same thing is an improvement ;-)

File: src/gwc/src/main/java/org/geoserver/gwc/FakeHttpServletRequest.java (hunk 2 above)
Comment (nmco) [reply src/gwc/src/main/java/org/geoserver/gwc/FakeHttpServletRequest.java]: Ok, fair enough.

File: src/gwc/src/test/java/org/geoserver/gwc/HttpRequestRecorderCallback.java (hunk 4 above)
Comment (aaime) [reply src/gwc/src/test/java/org/geoserver/gwc/HttpRequestRecorderCallback.java]: No problem

File: src/gwc/src/main/java/org/geoserver/gwc/FakeHttpServletRequest.java (hunk 1 above)
Comment (nmco) [reply src/gwc/src/main/java/org/geoserver/gwc/FakeHttpServletRequest.java]: :dancer:

**Pull request merged**